src/dmd/expressionsem.d

```diff
@@ -6228,6 +6229,37 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                 }
             }
         }
+        /* https://issues.dlang.org/show_bug.cgi?id=809
+         *
+         * If the address of a lazy variable is taken,
+         * the expression is rewritten so that the type
+         * of it is the delegate type. This means that
+         * the symbol is not going to represent a call
+         * to the delegate anymore, but rather, the
+         * actual symbol.
+         */
+        if (exp.e1.op == TOK.variable)
+        {
+            VarExp ve = cast(VarExp)exp.e1;
+            if (ve.var.storage_class & STC.lazy_)
+            {
+                exp.e1 = exp.e1.expressionSemantic(sc);
+                exp.e1 = resolveProperties(sc, exp.e1);
+                if (exp.e1.op == TOK.call)
+                {
+                    auto callExp = cast(CallExp)exp.e1;
+                    if (callExp.e1.type.ty == Tdelegate)
+                    {
+                        VarExp ve2 = cast(VarExp)callExp.e1;
+                        ve2.delegateWasExtracted = true;
```

Comment:
I do not understand why "extracting a delegate" permanently alters the semantics?

Reply:
Expression semantic on an AddrExp will rewrite the AST so that we have the name of a delegate instead of variable; after that when resolve properties is called on the newly rewritten AST node, when it finds the var name it will try to replace it again as a CallExp, so we need to store the fact that the delegate was already extracted.